src/engines/caching.cc

```diff
@@ -0,0 +1,315 @@
+       basePtr->All(context, callback);
+    }
+}
+
+int64_t CachingEngine::Count() {
+    LOG("Count");
+    string result;
+    int64_t noOfKeys = 0;
+
+    Each(&result, [](void* context, int kb, const char* k, int vb, const char* v){
+      const auto c = ((string*) context);
+      c->append(string(k, kb));
+      c->append(string(v, vb));
+    });
+
+    if (!result.empty())
+       noOfKeys = basePtr->Count();
+
+    return noOfKeys;
+}
+
+struct EachCacheCallbackContext {
+    KVEachCallback *cBack;
+    void* context;
+    std::list<std::string> *expiredKeys;
+};
+
+void CachingEngine::Each(void* context, KVEachCallback* callback) {
+    LOG("Each");
+    std::list<string> removingKeys;
+    EachCacheCallbackContext cxt = {callback, context, &removingKeys};
+
+    auto cb = [](void* context, int32_t kb, const char* k, int32_t vb, const char* v) {
+      const auto c = ((EachCacheCallbackContext*) context);
+      string localValue = v;
+      string timeStamp = localValue.substr(0,14);
+      string value = localValue.substr(14);
+      // TTL from config is ZERO or if the key is valid
+      if (!ttl || valueFieldConversion(timeStamp)) {
+      	(*c->cBack)(c->context, kb, k, value.length(), value.c_str());
+      } else {
+        c->expiredKeys->push_back(k);
+      } 
+    };
+
+    if (basePtr) {
+       basePtr->Each(&cxt, cb);
+       for (auto itr : removingKeys)
+          basePtr->Remove(itr);
+    }
+}
+
+KVStatus CachingEngine::Exists(const string& key) {
+    LOG("Exists for key=" << key);
+    string result;
+    api_flag = true;
+    KVStatus status = OK;
+
+    Get(&result, key, [](void* context, int vb, const char* v) {
+      const auto c = ((string*) context);
+      c->append(string(v, vb));
+    });
+
+    api_flag = false;
+    if (result.empty())
+       status = NOT_FOUND;
+    return status;
+}
+
+void CachingEngine::Get(void* context, const string& key, KVGetCallback* callback) {
+    LOG("Get key=" << key);
+
+    auto cb = [](void* context, int32_t vb, const char* v) {
+      const auto c = ((string*) context);
+      c->append(v, vb);
+    };
+
+    string value;
+    basePtr->Get(&value, key, cb);
+    bool timeValidFlag;
+    string valueField;
+    if (!value.empty()) {
+       string timeStamp = value.substr(0, 14);
+       valueField = value.substr(14);
+       timeValidFlag = valueFieldConversion(timeStamp);
+    }
+    //No value for a key on local cache or if TTL not equal to zero and TTL is expired
+    if (value.empty() || (ttl && !timeValidFlag)) {
+       // api_flag is true when Get api is called from Exists API
+       // api_flag is false when Get api is called from User application
+       if (api_flag == true || (remoteType == "Redis" && !getFromRemoteRedis(key, valueField)) || 
+	               (remoteType == "Memcached" && !getFromRemoteMemcached(key, valueField)) || 
+	               (remoteType != "Redis" && remoteType != "Memcached"))
+          return;
+    }
```

Comment:
This code should be refactored to avoid the `api_flag` -- better instead to repeat a little code or refactor out a common routine that does the time checking.  Current use of `api_flag` is not thread safe.